File: leetcode/0851-loud-and-rich/solution.py
from typing import List


class Solution:
    def loudAndRich(self, richer: List[List[int]], quiet: List[int]) -> List[int]:
        n = len(quiet)
        graph = [[] for _ in range(n)]
        for a, b in richer:
            graph[b].append(a)  # b is poorer than a, so explore from b towards richer a

        memo = [-1] * n

        def quietest(x: int) -> int:
            if memo[x] != -1:
                return memo[x]

            best = x
            for y in graph[x]:
                candidate = quietest(y)
                if quiet[candidate] < quiet[best]:
                    best = candidate

            memo[x] = best
            return best

        return [quietest(x) for x in range(n)]

File: leetcode/0851-loud-and-rich/Solution.java
import java.util.ArrayList;
import java.util.List;

class Solution {
    private List<List<Integer>> graph;
    private int[] quiet;
    private int[] memo;

    public int[] loudAndRich(int[][] richer, int[] quiet) {
        int n = quiet.length;
        this.quiet = quiet;
        graph = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            graph.add(new ArrayList<>());
        }
        for (int[] r : richer) {
            graph.get(r[1]).add(r[0]); // r[1] is poorer than r[0], explore from poorer towards richer
        }

        memo = new int[n];
        java.util.Arrays.fill(memo, -1);

        int[] answer = new int[n];
        for (int x = 0; x < n; x++) {
            answer[x] = quietest(x);
        }
        return answer;
    }

    private int quietest(int x) {
        if (memo[x] != -1) {
            return memo[x];
        }

        int best = x;
        for (int y : graph.get(x)) {
            int candidate = quietest(y);
            if (quiet[candidate] < quiet[best]) {
                best = candidate;
            }
        }

        memo[x] = best;
        return best;
    }
}
